### DISCUSSION TITLE
Adding raw signature values to `FeeMarketEIP1559Transaction` (or any other Transaction type)

### DISCUSSION THREAD

**Discussion by vdamle** [ethereumjs/ethereumjs-monorepo#3229] [General]
For apps that generate signatures from HSMs/hardware wallets etc., it would be nice if the library provides a way to input the ECDSA signature parameters generated by the signer and the y-parity, along with the unsigned TX payload and get the payload with signature.

I see there is a method that does exactly this but it is `protected`:

https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/tx/src/eip1559Transaction.ts#L311-L331

Obviously, I can copy-pasta that code, but it would be nice to expose that method. Thoughts?

**Comment by jochem-brouwer** [comment]
Does the HSM/hardware wallet return an `yParity` of either `0` or `1` for these signatures? (I assume you are using EIP1559 txs?)

**Reply by vdamle** [reply]
No, it does not. It returns only the `r` and `s` values. The `y-parity` is determined by attempting `ecrecover` on both values and using the value that results in a successfully recovery of the public key. If the SDK can take care of that as well, even better :-)

**Reply by vdamle** [reply]
Re: your other point, yes I'm using EIP1559 tx right now, but the same applies to other tx types as well.

**Reply by jochem-brouwer** [reply]
I am not really into this HSM world, could you explain how this works? So you sign a message, you get the s,r values and have a known sender and now try to derive the y parity by using ecrecover twice on the two different partities? Checking for the "correct" sender?

If this is the case you now have all values to directly construct a transaction via `fromTxData`. (Unless i am missing something)

**Reply by holgerd77** [reply]
Any follow-up here?

**Reply by vdamle** [reply]
Sorry for the delay. You are correct that I _can_ get the correct `r` and `s` and `v` values from the signature returned from the HSM/any external wallet -- after a little bit of gymastics -- specifically, changing `s` value to make sure it is non-malleable and also determine the correct `v`.

* So I started by instantiating a Transaction with a call to `fromTxData` where I supply all relevant values as params.
* I then generate the 32 byte hash that I can send to an external signing service.
* After I have the signature parameters, I have to call `fromTxData` again by supplying all relevant params that now includes the signature params (duplication, IMO)

To me, that feels like extra boilerplate because:

* `r`, `s` and `v` are read only variables.
* there is no helper function in the library for "Look, i created this transaction without signature params. I now have the signature params, could you please add these to the transaction and give me back what I need to submit to the node?"

That said, I do agree that an end user can the same code that `_processSignature()` has accomplish the above.

**Reply by jochem-brouwer** [reply]
Aha, now I indeed see the redundancy which is indeed somewhat unnecessary. What value does your HSM return for `v`? Is it `0` or `1` or something else?

We should indeed expose `_processSignature` or something similar (maybe overload `sign` to accept `r`, `s`, `v` values?). We will discuss :smile:

**Reply by vdamle** [reply]
Thanks @jochem-brouwer - ether of those solutions would work just fine! HSM only returns `{r, s}` for the signature. The `v` bit is computed locally by attempting recovery for `v = {0, 1}` and use whichever value results in a public key/address match.

**Reply by jochem-brouwer** [reply]
I addressed this today, I think this would work for you? https://github.com/ethereumjs/ethereumjs-monorepo/pull/3238

**Reply by vdamle** [reply]
The API looks great @jochem-brouwer , thanks for the enhancement! 🙏🏼
